### PR TITLE
Add snapshot account state verify proposal

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem
 
xor method of rolling up hashes of live account states is too easily attacked by probabilistic collision.

#### Summary of Changes

Add proposal to add account states to two sets of accumulators to be able to validate the account set in a snapshot matches what is on the network.

Fixes #

### CODE REVIEW

File: book/src/proposals/snapshot-verification.md [hunk 1]
```diff
@@ -2,8 +2,41 @@
...
+attacker could give the validator an incorrect state, and then try to convince it to accept a transaction that would otherwise be rejected.
 
 ## Solution
 
-While a validator is processing transactions to catch up to the cluster from the snapshot, use incoming vote transactions and the commitment calculator to confirm that the cluster is indeed building on the snapshotted bank hash. Once a threshold commitment level is reached, accept the snapshot as valid and start voting.
+Currently, the bank hash is derived from xor'ing an image derived from the hash of the
+account state and slot. Unfortunately this is insecure as an xor collision can be found
+with O(n) bits in the image and O(n^4) computation.
+
+We propose to use an RSA accumulator where we hash the account state to a large prime-ish number.
```
Comment (aeyakovenko) [inline book/src/proposals/snapshot-verification.md]: Write it as if the proposal has been accepted.  We don’t need the old one referenced.

File: book/src/proposals/snapshot-verification.md [hunk 2]
```diff
@@ -2,8 +2,41 @@
...
+* Account owner
+* Account data
+* Account pubkey
+* Account lamports balance
+* Fork the account is stored on
+
+That hash is added to an RSA accumulator in standard `g^(hash) mod n` way.
+
+Since adding to an accumulator is a single-threaded process, multiple accumulators can
+be used to obtain parallelism for the entire account set in the slot. They can be indexed some
```
Comment (aeyakovenko) [inline book/src/proposals/snapshot-verification.md]: Can you cover this?  Can we index by hash of the account to force rotation?

Comment (sakridge) [reply book/src/proposals/snapshot-verification.md]: Is there an attack there? or it just seems more careful?

You can do some kind of dos if you can force a validator to calculate a large amount of updates to a single generator, but it seems like you could grind on the hash of the account state as well.

Comment (aeyakovenko) [reply book/src/proposals/snapshot-verification.md]: @sakridge yep.  I think there is a dos either way.  but if we force redistribution on write it should rebalance things.

Comment (sakridge) [reply book/src/proposals/snapshot-verification.md]: We could sprinkle in some of the previous block hash bits, so an attacker only has a limited time to grind.

File: book/src/proposals/snapshot-verification.md [hunk 3]
```diff
@@ -2,8 +2,37 @@
 
 ## Problem
 
-Snapshot verification of the account states is implemented, but the bank hash of the snapshot which is used to verify is falsifiable.
+When a validator boots up from a snapshot, it needs a way to verify the account set matches what the rest of the network sees quickly. A potential
```
Comment (t-nelson) [inline book/src/proposals/snapshot-verification.md]: nit: wrap at 80

File: book/src/proposals/snapshot-verification.md [hunk 4]
```diff
@@ -2,8 +2,37 @@
...
+* Account owner
+* Account data
+* Account pubkey
+* Account lamports balance
+* Fork the account is stored on
+
+That hash is added to an RSA accumulator in standard `g^(hash) mod n` way.
+
+Since adding to an accumulator is a single-threaded process, multiple accumulators can
+be used to obtain parallelism for the entire account set in the slot. They can be indexed some
```
Comment (t-nelson) [inline book/src/proposals/snapshot-verification.md]: nit: "indexed **by** some"

File: book/src/proposals/snapshot-verification.md [hunk 5]
```diff
@@ -2,8 +2,37 @@
...
+Propose to use an RSA accumulator where it hashes the account state to a large prime-ish number
+using a primality test and incrementing a nonce to find a hash that passes the primality test.
+
+On account store of non-zero lamport accounts, we hash the following data:
+
+* Account owner
+* Account data
+* Account pubkey
+* Account lamports balance
+* Fork the account is stored on
```
Comment (ryoqun) [inline book/src/proposals/snapshot-verification.md]: nits: `Slot` (rename remnant?)

Comment (ryoqun) [reply book/src/proposals/snapshot-verification.md]: nits: Also these days, hash includes `rent_epoch` and `executable` #7415

**Pull request closed without merging**